crates/oxide/src/extractor/candidate_machine.rs

```diff
@@ -0,0 +1,378 @@
+                        // not preceded by another variant, and followed by any kind of whitespace
+                        // or the end of the line.
+                        //
+                        // E.g.: `{ underline: true }`
+                        //          ^^^^^^^^^^        Variant
+                        //          ^^^^^^^^^         Utility (followed by `: `)
+                        let after = cursor.input.get(utility_span.end + 2).unwrap_or(&b'\0');
+                        if after.is_ascii_whitespace() || *after == b'\0' {
+                            cursor.move_to(utility_cursor.pos + 2);
+                            return self.done_span(utility_span);
+                        }
+
+                        self.start_pos = variant_span.start;
+                    }
+
+                    //
+                    self.last_variant_end_pos = Some(variant_cursor.pos);
+                }
+
+                // Variant is done, utility is invalid
+                (MachineState::Done(variant_span), MachineState::Idle) => {
+                    cursor.move_to(variant_cursor.pos + 1);
+
+                    if let Some(end_pos) = self.last_variant_end_pos {
+                        if end_pos + 1 > variant_span.start {
+                            self.reset();
+                            return MachineState::Idle;
+                        }
+                    } else {
+                        self.start_pos = variant_span.start;
+                    }
+
+                    self.last_variant_end_pos = Some(variant_cursor.pos);
+                }
+
+                (MachineState::Idle, MachineState::Idle) => {
+                    // Skip main cursor to the next character after both machines. We already know
+                    // there is no candidate here.
+                    if variant_cursor.pos > cursor.pos || utility_cursor.pos > cursor.pos {
+                        cursor.move_to(variant_cursor.pos.max(utility_cursor.pos));
+                    }
+
+                    self.reset();
+                    cursor.advance();
+                }
+            }
+        }
+
+        MachineState::Idle
+    }
+}
+
+impl CandidateMachine {
+    #[inline(always)]
+    fn done_span(&mut self, span: Span) -> MachineState {
+        self.reset();
+        MachineState::Done(span)
+    }
+}
+
+/// A candidate must be preceded or followed by any of these characters
+/// E.g.: `<div class="flex">`
+///                   ^ Valid for `flex`
+///        ^ Invalid for `div`
+#[inline(always)]
+fn is_valid_common_boundary(c: &u8) -> bool {
+    matches!(
+        c,
+        b'\t' | b'\n' | b'\x0C' | b'\r' | b' ' | b'"' | b'\'' | b'`' | b'\0'
+    )
+}
+
+/// A candidate must be preceded by any of these characters.
+/// E.g.: `%w[flex]`
+///          ^
+#[inline(always)]
+fn is_valid_before_boundary(c: &u8) -> bool {
+    is_valid_common_boundary(c) || matches!(c, b'{' | b'[' | b'.')
```

Comment:
This should probably include `(` for `%w(…)` Ruby syntax

Reply:
Not necessary anymore because of the pre-processing step. 🥳